### ISSUE TITLE
High memory usage

### ISSUE THREAD

**Issue by mistlebee** [FXMisc/RichTextFX#879]
## Expected Behavior

I want to use the assist field like an editor with an assist function. But if the text field has a long lifetime, the memory increases a lot.
I would expect that memory usage falls, if I delete content from field.
I would also expect, that I have a similar memory usage whether I insert the content step by step or insert the whole content via copy/paste. (if I insert a lot of content via copy/paste the memory usage is very low).

## Actual Behavior

I tried to profile the field. Here are my results:

1. Memory usage is low at the beginning:
![2019-12-04 14_48_18-Java VisualVM](https://user-images.githubusercontent.com/26139406/70149839-399b8900-16a9-11ea-8b44-f99bc6b18191.png)
2. Start writing text with stylings and create an initial heap dump (after performing garbage collection)
![2019-12-04 14_48_51-Richt Text Test](https://user-images.githubusercontent.com/26139406/70149840-399b8900-16a9-11ea-9e8c-079b04e771b0.png)
3. Write more text with stylings
![2019-12-04 14_52_06-Richt Text Test](https://user-images.githubusercontent.com/26139406/70149841-3a341f80-16a9-11ea-8006-4f54c4a15ff3.png)
3. After manual triggering of garbage collector. Memory is very high
![2019-12-04 14_52_56-Java VisualVM](https://user-images.githubusercontent.com/26139406/70149842-3a341f80-16a9-11ea-887c-a651702d5522.png)
4. New heap dump in comparison with the last one
![2019-12-04 14_55_52-Java VisualVM](https://user-images.githubusercontent.com/26139406/70149843-3a341f80-16a9-11ea-8530-900bbf463e19.png)
5. Clear text
![2019-12-04 15_01_54-Richt Text Test](https://user-images.githubusercontent.com/26139406/70149845-3accb600-16a9-11ea-9653-8a155cd53d11.png)
6. Create new heap dump after garbage collection and compare to last one
![2019-12-04 15_08_06-Java VisualVM](https://user-images.githubusercontent.com/26139406/70149846-3accb600-16a9-11ea-9c63-4d572bcb4209.png)
7. Memory usage remains high
![2019-12-04 15_08_38-Java VisualVM](https://user-images.githubusercontent.com/26139406/70149848-3b654c80-16a9-11ea-8554-b1d24ffb51d9.png)

With each change of text field I wand to reparse the field and update the stylings. But it seems like old stylings are never cleaned up.
I tried to use another Undomanager. Because I just need text-changes in UndoManager. I can recreate the stylings with the text parsing. But even with such a simple UndoManger the memory increases a lot.

## Reproducible Demo
Here is my demo code.
[richtexttest.zip](https://github.com/FXMisc/RichTextFX/files/3922273/richtexttest.zip)

## Environment info:

- RichTextFX Version: 0.10.2
- Operating System: Windows 10
- Java version:1.8.121

<hr>

## Current Workarounds

It's possible to use dispose. But during life time of the component it is hard to find a good trigger point for recreation. Also redo/undo is lost with this workaround.

**Comment by Groostav** [comment]
I feel like:

1. this must be a show stopper for you. That ram usuage is totally unacceptable
2. On both jdk 1.8.222, and java 13 _I dont see this problem_
3. this might be related to a redundant CSS bug that has been fixed, reverted, and fixed again in openjfx: https://bugs.openjdk.java.net/browse/JDK-8193445

**Comment by mistlebee** [comment]
Thank you very much for your response!
I can reproduce the problem with 1.8.231, too.
I just start the demo application and copy/paste the string **FAM1:Item11** (comma separated) again and again.

Do you cannot reproduce the problem with the same the application?
Maybe there is a mistake in my code and I should use another method to set the styles?

**Comment by Groostav** [comment]
Yeah, so I ran your project from intelliJ using java 8 u 222 (azul's zulu java 8.40.0.25):

![image](https://user-images.githubusercontent.com/1680611/70673967-46842380-1c39-11ea-8cbe-52caaa0a59bf.png)

(AFAIK azul doesn't add anything to their java builds, so I don't think its something special about hotspot vs zulu).

**Comment by Jugen** [comment]
I've finally started looking into this and thought it would be helpful to establish a baseline and take it from there.

So I created three tests using `TextArea`, `TextFlow`, and RichTextFX's `StyleClassedTextArea` (with a _No-op_ undo manager). Then I pasted **FAM1:Item11** two hundred times into each and noted their memory usages. (The latter two styled any pasted text on the fly, without reprocessing previously pasted text.)

I ran the test on all three multiple times with different JRE's (8.0_101, 10.0.2, and 11.0.4 32bit). I used  Java's `Runtime.totalMemory - Runtime.freeMemory` and Window's TaskManager to get usage stats, with the average results as follows:

Text Control | Java Mem | Task Man
------------|------------|----------
TextArea | ~20 MB | ~95 MB
TextFlow | ~50 MB | ~120 MB
RichTextFX | ~200 MB | ~360 MB

The above clearly shows that the RichTextFX test uses **3** to **4** times more memory than the TextFlow test!

One of the biggest differences between these two test cases is that my TextFlowTest uses only a single TextFlow instance whereas RichText recreates the TextFlow control everytime there's a change. So I modified RichTextFX to reuse the TextFlow instead with the following result:

Text Control | Java Mem | Task Man
------------|------------|----------
RichTextFX | ~25 MB | ~130 MB

As can be seen this much more inline with what it should be I think. The problem is that using this configuration causes RichTextFX to fail 21 of it's test cases. I'll be looking more into this in the next few weeks and hopefully find a solution .....
